lib/ruby_lsp/ruby_lsp_rails/runner_client.rb

```diff
@@ -0,0 +1,77 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "json"
+require "open3"
+
+# NOTE: We should avoid printing to stderr since it causes problems. We never read the standard error pipe
+# from the client, so it will become full and eventually hang or crash.
+# Instead, return a response with an `error` key.
+
+module RubyLsp
+  module Rails
+    class RunnerClient
+      extend T::Sig
+
+      sig { void }
+      def initialize
+        stdin, stdout, stderr, wait_thread = Open3.popen3(
+          "bin/rails",
+          "runner",
+          "#{__dir__}/server_start.rb",
+        )
+        @stdin = T.let(stdin, IO)
+        @stdout = T.let(stdout, IO)
+        @stderr = T.let(stderr, IO)
+        @wait_thread = T.let(wait_thread, Process::Waiter)
+        @stdin.binmode # for Windows compatibility
+        @stdout.binmode # for Windows compatibility
+      end
+
+      sig { params(name: String).returns(T.nilable(T::Hash[Symbol, T.untyped])) }
+      def model(name)
+        make_request("model", name: name)
+      end
+
+      sig { void }
+      def shutdown
+        send_notification("shutdown")
+        sleep(0.01) while @wait_thread.alive?
+        [@stdin, @stdout, @stderr].each(&:close)
+      end
+
+      private
+
+      sig { params(request: T.untyped, params: T.untyped).returns(T.untyped) }
+      def make_request(request, params = nil)
+        send_message(request, params)
+        read_response
+      end
+
+      sig { params(request: T.untyped, params: T.untyped).void }
+      def send_message(request, params = nil)
+        json = { method: request, params: params }.compact.to_json
```

Comment:
I think compact might create another hash in between.
```suggestion
        message = { method: request }
        message[:params] = params if params
        json = message.to_json
```